src/runtime-rs/crates/hypervisor/src/device/topology.rs

```diff
@@ -0,0 +1,240 @@
+
+Plug only PCI Express devices into PCI Express Ports.
+pcie.0 bus (Root Complex)
+----------------------------------------------------------------------------
+|    |    |    |    |    |    |    |    |    |    |    |    |    |    | .. |
+------|----------------|--------------------------------------|-------------
+      |                |                                      |
+      V                V                                      V
+ -------------     -------------                        -------------
+ | Root Port |     | Root Port |                        | Root Port |
+ -------------     -------------                        -------------
+     |                                                       |
+     |                              -------------------------|-----------------------
+------------                        |                -----------------              |
+| PCIe Dev |                        |   PCI Express  | Upstream Port |              |
+------------                        |     Switch     -----------------              |
+                                    |                 |            |                |
+                                    |   -------------------    -------------------  |
+                                    |   | Downstream Port |    | Downstream Port |  |
+                                    |   -------------------    -------------------  |
+                                    -------------|-----------------------|-----------
+                                            ------------
+                                            | PCIe Dev |
+                                            ------------
+*/
+
+use std::collections::HashMap;
+
+use crate::device::pci_path::PciSlot;
+use anyhow::Result;
+use kata_types::config::hypervisor::TopologyConfigInfo;
+
+use super::pci_path::PciPath;
+use crate::HYPERVISOR_DRAGONBALL;
+
+pub const PCIE_ROOT_BUS_SLOTS_CAPACITY: u32 = 32;
+pub const PCIE_TOPOLOGY_SLOTS_CAPACITY: u32 = 32;
+const DEFAULT_PCIE_ROOT_BUS: &str = "pcie.0";
+const DEFAULT_PCIE_ROOT_BUS_ADDRESS: &str = "0000:00";
+
+pub trait PCIeDevice: Send + Sync {
+    fn device_type(&self) -> PCIeDevType;
+    fn device_id(&self) -> &str;
+}
+
+// PCI-PCI Bridge
+pub struct PCIBridge {
+    _pci_endpoints: HashMap<u32, PCIeEndpoint>,
+}
+
+#[derive(Default)]
+pub struct PCIePCIBridge {
+    pub root_port: PCIeRootPort,
+    _bridge_devices: HashMap<String, PCIeEndpoint>,
+}
+
+// PCIe Root Complex
+#[derive(Clone, Debug, Default)]
+pub struct PCIeRootComplex {
+    pub root_bus: String,
+    pub root_bus_address: String,
+    pub root_bus_devices: HashMap<String, PCIeEndpoint>,
+}
+
+// PCIe Root Port
+#[derive(Clone, Debug, Default)]
+pub struct PCIeRootPort {
+    pub chassis: u8,
+    pub slot: u8,
+}
+
+#[derive(Default)]
+pub struct UpStreamPort {
+    pub root_port: PCIeRootPort,
+    pub bus_numer: u8,
+    pub port_id: String,
+}
+
+#[derive(Default)]
+pub struct DownStreamPort {
```

Comment:
The same comment here: Is it going to be possible to set mem-reserve and pref-64-reserve for the root port? Does dragonball has such settings for the root port? The GPU has "huge" BARs and they need proper IOMMU memory setup. We cannot work with the default 4k window range.

Reply:
The fields related to resource reserved will be added.